### PR TITLE
feat: Expose local storage through access control facility

### LINKED ISSUE #3180: Allow users to access persistent data from Zowe Explorer local storage
**Is your feature request related to a problem? Please describe.**

In v2, persistent data like filter history and favorites for data sets, USS, and jobs were stored in the Visual Studio Code settings. This made it simple for extenders to access and read this data.
In v3, this data was moved to VS Code storage, and now there is no way for extenders to access it.

**Describe the solution you'd like**

I would like this persistent data to be accessible for extenders like it was in v2. This could be in the form of registered commands for getting/setting persistent data. You could also expose each resource's `ZoweTreeProvider to extenders.

**Describe alternatives you've considered**

I tried accessing the ZE's new storage using the keys available in the Zowe Explorer API, but because I cannot access Zowe Explorer's extension context, there is no way for me to access its memento storage. 

**Additional context**

### PR DESCRIPTION
**Pull request by traeok**
## Proposed changes

Resolves #3180 

**Dev doc:** https://github.com/zowe/zowe-explorer-vscode/wiki/Local-Storage

### How to test

- Build a VSIX for Zowe Explorer using this PR and install it in VS Code
- Checkout the `local-storage-sample` branch on https://github.com/zowe/zowe-client-samples
- `cd vscode-extension-samples/local-storage-sample && npm install && npm run compile`
- After the sample build has finished, click on the Run & Debug panel on the VS Code Side Bar
- Select "Run Extension" from the tasks list at the top and select the Play ▶ button
- Notice the notifications for the sample extension:
  - The first notification shown is for the list of writable keys
  - 2nd notification: list of readable keys
  - 3rd notification: the value for `zowe.ds.history` (a readable key)
  - 4th notification: trying to access a key that isn't exposed

## Release Notes

Milestone: 3.1.0

Changelog:

- Exposed read and write access to local storage keys for Zowe Explorer extenders. [#3180](https://github.com/zowe/zowe-explorer-vscode/issues/3180)

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
  - Fixed issue where workspace settings were merged into wrong local storage layer during migration
- [x] Enhancement (non-breaking change which adds or improves functionality)
  - Added support for `ZoweLocalStorage` to save in a workspace
  - Added the local storage access facility for extenders
- [ ] Breaking change (a change that would cause existing functionality to not work as expected)
- [ ] Documentation (Markdown, README updates)
- [ ] Other (please specify above in "Proposed changes" section)

## Checklist

**General**

- [x] I have read the [CONTRIBUTOR GUIDANCE](https://github.com/zowe/zowe-explorer-vscode/wiki/Best-Practices:-Contributor-Guidance) wiki
- [x] All PR dependencies have been merged and published (if applicable)
- [ ] A GIF or screenshot is included in the PR for visual changes
- [x] The pre-publish command has been executed:
  - **v2 and below:** `yarn workspace vscode-extension-for-zowe vscode:prepublish`
  - **v3:** `pnpm --filter vscode-extension-for-zowe vscode:prepublish`

**Code coverage**

- [x] There is coverage for the code that I have added
- [x] I have added new test cases and they are passing
- [x] I have manually tested the changes

**Deployment**

- [x] I have added developer documentation (if applicable)
- [ ] Documentation should be added to Zowe Docs
  - If you're an outside contributor, please post in the [#zowe-doc Slack channel](https://openmainframeproject.slack.com/archives/CC961JYMQ) to coordinate documentation.
  - Otherwise, please check with the rest of the squad about any needed documentation before merging.
- [ ] These changes may need ported to the appropriate branches (list here):

### CODE REVIEW

**Review by anaxceron — DISMISSED** [review]

**Review by JillieBeanSim — CHANGES_REQUESTED** [review]
thanks for working on this @traeok 
I do see that it wiped my zowe history which may need investigating, we wouldn't want to have that happen to customers.  I didn't see the pop up described for the sample, I received the migration pop up with restart button.

not sure if someone else can replicate the issue I had seen or if just with me.

**Comment by traeok** [comment]
> thanks for working on this @traeok I do see that it wiped my zowe history which may need investigating, we wouldn't want to have that happen to customers. I didn't see the pop up described for the sample, I received the migration pop up with restart button.
> 
> not sure if someone else can replicate the issue I had seen or if just with me.

Thanks @JillieBeanSim for catching this, I realized that we need to use a `defaultValue` of undefined when accessing `workspaceState` now that our local storage class supports it. Addressed in bd10595

**Review by anaxceron — APPROVED** [review]

**Review by zFernand0 — DISMISSED** [review]
TL;DR:  LGTM! 😋 
Sorry for the long text below 😅 

---

Question, should we merge the `local-storage-sample` branch into [zowe-client-samples # main](https://github.com/zowe/zowe-client-samples)?
Also, should there be an activationEvent (maybe `onStartupFinished`)? Without it, I didn't get any notifications since the local-storage sample extension wouldn't activate 😋 

Here is what I got when I added the activationEvent
image

And as expected, the notifications are in reverse order 😋 

---

It seems like could benefit from running prettier on this branch 😋 
```
modified:   packages/zowe-explorer/__tests__/__unit__/trees/dataset/DatasetFSProvider.unit.test.ts
modified:   packages/zowe-explorer/l10n/bundle.l10n.json
modified:   packages/zowe-explorer/l10n/poeditor.json
```

---

For those testing, please make sure that your .npmrc isn't pointing the `@zowe` scoped registry to something other than the public NPM registry 😅 
```
npm error code E404
npm error 404 Not Found - GET https://zowe.jfrog.io/zowe/api/npm/npm-local-release/@zowe%2fzowe-explorer-api
npm error 404
npm error 404  '@zowe/zowe-explorer-api@^3.0.0' is not in this registry.
npm error 404
npm error 404 Note that you can also install from a
npm error 404 tarball, folder, http url, or git url.
```

**Review by zFernand0 — APPROVED** [review]

**Review by zFernand0 — APPROVED** [review]

**Review by anaxceron — APPROVED** [review]

**Comment by traeok** [comment]
Sorry for the extra mentions, I've disabled the merge by workflow because I need to make updates to the actions that are used, as well as fixing an issue with duplicate comments for the 24hour reminder.

**Review by zFernand0 — APPROVED** [review]
I thought I had approved this before 😋 
LGTM! 😋

**Review by awharn — APPROVED** [review]
